SciLean/Data/Mesh/SurfaceMesh.lean

```diff
@@ -0,0 +1,350 @@
+/-
+Copyright (c) 2023 Siddharth Bhat. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat
+
+Surface Mesh based on discrete differential geometry.
+https://github.com/GeometryCollective/geometry-processing-js/blob/c7ea47ae808979a5361e3fcd61bf921a194bf897/core/mesh.js#L3-L213
+-/
+
+import SciLean.Core
+
+
+-- abbrev V3 : Type := ℝ^{3}
+-- 3D vector.
+structure V3 where
+  x : ℝ
+  y : ℝ
+  z : ℝ
+
+/- index of vertices --/
+abbrev Index (name : Name) := Nat
+abbrev Index.invalid : Index α := 424242
+
```

Comment:
Interesting choice of 'typing' the indices. I would probably have a small struct containing mesh's element counts and the have indices 'typed' by this struct. 

This way you guarantee in bound accesses, but probably run into difficulties with dependent types once you start writing functions adding and removing indices.

What are your thoughts on providing type safety for indices?

Reply:
@lecopivo I would love to use dependent types, but I am generally too scared of the indexing annoyances to bother with them. I wish there was some way type checked indexes could be 'opt in'.

Reply:
for example, we play tricks like first initializing all the vertices, and then random accessing into the now-initialized vertex array. These would be annoying to make dependent. A good motivation to make them dependent is to eventually write a paper on this stuff, where having the dependent types is useful for brownie points :)

Reply:
"good motivation to make them dependent is to eventually write a paper on this stuff"

Nope, that is a bad motivation :) a good motivation is that it might produce a nice to use API but I get your point ;)